clusterlib/futures.py

```diff
@@ -0,0 +1,638 @@
+"""Implementation of the concurrent.futures API for a cluster
+
+See:
+
+https://docs.python.org/3/library/concurrent.futures.html
+
+"""
+# Authors: Olivier Grisel
+#
+# License: BSD 3 clause
+from __future__ import print_function
+from getpass import getuser
+import os
+import os.path as op
+from time import time, sleep
+import sys
+import logging
+import joblib
+import signal
+import errno
+
+from clusterlib.scheduler import submit, queued_or_running_jobs
+
+logger = logging.getLogger('clusterlib')
+TRACE = logging.DEBUG - 5  # Even more verbose than DEBUG
+
+
+CANCELLED = 'cancelled'
+RUNNING = 'running'
+FINISHED = 'finished'
+PENDING = 'pending'
+DEFAULT_POLL_INTERVAL = 10  # in seconds
+CANCELLATION_SIGNALS = [
+    signal.SIGTERM,
+    signal.SIGABRT,
+    signal.SIGQUIT,
+]
+MAX_FS_CACHE_DELAY = 100
+
+
+try:
+    from concurrent.futures import TimeoutError, CancelledError
+
+    class ClusterTimeoutError(TimeoutError):
+        pass
+
+    class ClusterCancelledError(CancelledError):
+        pass
+
+except ImportError:
+    # define our own exception classes to keep the dependency
+    # on the "futures" package optional under Python 2.
+    # concurrent.futures is shipped by default in the Python 3
+    # standard library
+    class ClusterTimeoutError(Exception):
+        pass
+
+    class ClusterCancelledError(Exception):
+        pass
+
+
+def safe_makedirs(folder):
+    """Protect makedirs for concurrent."""
+    try:
+        os.makedirs(folder)
+    except OSError:
+        if not op.exists(folder):
+            raise
+        # else: folder already exists: ignore
+
+
+def _dump(obj, filename):
+    joblib.dump(obj, filename)
+
+
+def _load(filename, mmap_mode=None):
+    return joblib.load(filename, mmap_mode=mmap_mode)
+
+
+class AtomicMarkerError(Exception):
+    pass
+
+
+class AtomicMarker(object):
+    """Named marker for coordination between concurrent processes
+
+    clusterlib leverages a shared filesystem to exchange data between
+    the main coordination script and worker processes.
+
+    Such filesystem have typically very few atomic operations. This
```

Comment:
Out of curiosity, what kind of shared filesystem(s) do you have on your cluster(s) NFS? FHGFS? Lustre? GlusterFS? CephFS? something else? Some filesystems can use NFS clients (e.g. GlusterFS) so the caching / locking issues might still happen with those.